app/src/main/java/ai/elimu/appstore/synchronization/AppListAdapter.java

```diff
@@ -121,14 +132,14 @@ public void onBindViewHolder(final ViewHolder holder, final int position) {
                 locale = Locale.EN;
             }
             String language = locale.getLanguage();
-            String fileName = applicationVersion.getApplication().getPackageName() + "-" +
+            final String fileName = applicationVersion.getApplication().getPackageName() + "-" +
                     applicationVersion.getVersionCode() + ".apk";
-            File apkDirectory = new File(Environment.getExternalStorageDirectory() + "/" +
+            final File apkDirectory = new File(Environment.getExternalStorageDirectory() + "/" +
                     ".elimu-ai/appstore/apks/" + language);
             final File existingApkFile = new File(apkDirectory, fileName);
             Timber.i("existingApkFile: " + existingApkFile);
             Timber.i("existingApkFile.exists(): " + existingApkFile.exists());
-            if (existingApkFile.exists()) {
+            if (isExistingAndValidApk(existingApkFile, applicationVersion.getChecksumMd5())) {
```

Comment:
Do not calculate the checksum of each while when scrolling through the list. This causes really laggy UI.